flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/RecordWriterOutput.java

```diff
@@ -101,32 +97,51 @@ public void collect(StreamRecord<OUT> record) {
     }
 
     private <X> void pushToRecordWriter(StreamRecord<X> record) {
-        serializationDelegate.setInstance(record);
+        if (announcedStatus.isIdle()) {
```

Comment:
I think we need to add a comment here to explain why this can happen.

Reply:
The comment is good in respect to explaining the invariant (you could also give a link to the docs). However, it's not helping the reader to understand why this case is happening.